File: suite/Primes/Primes.rs
use std::rc::Rc;

#[derive(Debug, Clone)]
enum List<A> {
    Nil,
    Cons(A, Rc<List<A>>),
}

impl List<i64> {
    fn len(&self) -> i64 {
        match self {
            List::Nil => 0,
            List::Cons(_, as_) => 1 + as_.len(),
        }
    }

    fn interval(m: i64, n: i64) -> List<i64> {
        if n < m {
            List::Nil
        } else {
            List::Cons(m, Rc::new(List::interval(m + 1, n)))
        }
    }
}

fn remove_multiples(n: i64, l: List<i64>) -> List<i64> {
    match l {
        List::Nil => List::Nil,
        List::Cons(x, xs) => {
            if x % n == 0 {
                remove_multiples(n, Rc::unwrap_or_clone(xs))
            } else {
                List::Cons(x, Rc::new(remove_multiples(n, Rc::unwrap_or_clone(xs))))
            }
        }
    }
}

fn sieve(l: List<i64>) -> List<i64> {
    match l {
        List::Nil => List::Nil,
        List::Cons(x, xs) => List::Cons(
            x,
            Rc::new(sieve(remove_multiples(x, Rc::unwrap_or_clone(xs)))),
        ),
    }
}

fn main_loop(iters: u64, n: i64) {
    let res = sieve(List::interval(2, n));
    if iters == 1 {
        println!("{}", res.len());
    } else {
        main_loop(iters - 1, n)
    }
}

fn main() {
    let mut args = std::env::args();
    args.next();
    let iters = args
        .next()
        .expect("Missing Argument iterations")
        .parse::<u64>()
        .expect("Iterations must be a number");
    let n = args
        .next()
        .expect("Missing Argument n")
        .parse::<i64>()
        .expect("n must be a number");
    main_loop(iters, n)
}

File: suite/Primes/Primes.ml
let rec len l =
  match l with
    | [] -> 0
    | _ :: xs -> 1 + (len xs)

let rec interval_list m n =
  if n < m then []
  else m :: interval_list (m + 1) n

let rec remove_multiples n l =
  match l with
    | [] -> []
    | x :: xs ->
        if x mod n = 0
        then remove_multiples n xs
        else x :: remove_multiples n xs

let rec sieve l =
  match l with
    | [] -> []
    | x :: xs -> x :: sieve (remove_multiples x xs)

let rec main_loop iters n =
  let res = sieve (interval_list 2 n) in
  if iters = 1 then
    print_endline (string_of_int (len res))
  else main_loop (iters - 1) n

let main =
  let iters = int_of_string Sys.argv.(1) in
  let n = int_of_string Sys.argv.(2) in
  main_loop iters n
